### PR TITLE
gh-119352 Add UOp Pair Compatibility to PyStats

### LINKED ISSUE #119352: Add UOp Pair Compatibility to PyStats
When it comes time to start implementing superinstructions for the Tier 2 interpreter, a reasonable place to start will be choosing pairs of UOps that use disjoint sets of the possible inputs (oparg, operand, and target). This way, the superinstructions can be handled just like the existing UOps, without having to cram additional opargs/operands/targets in for each UOp of the superinstruction.

To this end, having this information contained in the output of a PyStats run (and `summarize_stats.py`) will be helpful. 

### Linked PRs
* gh-119354

### PR DESCRIPTION
**Pull request by JeffersGlass**
This PR adds functionality to `PyStats`-enabled builds, specifically to the UOp pair-count section emitted when the interpreter is build with the Tier 2 Interpreter. It adds another column to the pair-count table, indicating whether the two instructions have overlapping uses of `oparg`, `operand`, and `target`. This makes it easy to tell which pairs can be immediately adapted to become superinstructions, and which would require a modified instruction format.

See below for an example of this output - the `Compatibility` column in the far right is the added column.

Closes #119352

<table>
    <thead>
    <tr>
    <th align="left">Pair</th>
    <th align="right">Count</th>
    <th align="right">Self</th>
    <th align="right">Cumulative</th>
    <th align="left">Compatibility</th>
    </tr>
    </thead>
    <tbody>
    <tr>
    <td align="left">_LOAD_FAST_0 _GUARD_TYPE_VERSION</td>
    <td align="right">62,941,340</td>
    <td align="right">7.6%</td>
    <td align="right">7.6%</td>
    <td align="left">No Conflicts</td>
    </tr>
    <tr>
    <td align="left">_CHECK_MANAGED_OBJECT_HAS_VALUES _LOAD_ATTR_INSTANCE_VALUE_0</td>
    <td align="right">37,435,120</td>
    <td align="right">4.5%</td>
    <td align="right">12.2%</td>
    <td align="left">Conflict. Both Use: Target</td>
    </tr>
    <tr>
    <td align="left">_GUARD_TYPE_VERSION _CHECK_MANAGED_OBJECT_HAS_VALUES</td>
    <td align="right">37,435,120</td>
    <td align="right">4.5%</td>
    <td align="right">16.7%</td>
    <td align="left">Conflict. Both Use: Target</td>
    </tr>
    <tr>
    <td align="left">_COPY _TO_BOOL_BOOL</td>
    <td align="right">25,148,840</td>
    <td align="right">3.0%</td>
    <td align="right">19.8%</td>
    <td align="left">No Conflicts</td>
    </tr>
    <tr>
    <td align="left">_GUARD_TYPE_VERSION _EXIT_TRACE</td>
    <td align="right">24,884,600</td>
    <td align="right">3.0%</td>
    <td align="right">22.8%</td>
    <td align="left">No Conflicts</td>
    </tr>
    <tr>
    <td align="left">_PUSH_FRAME _RESUME_CHECK</td>
    <td align="right">23,560,800</td>
    <td align="right">2.9%</td>
    <td align="right">25.6%</td>
    <td align="left">No Conflicts</td>
    </tr>
    <tr>
    <td align="left">_SAVE_RETURN_OFFSET _PUSH_FRAME</td>
    <td align="right">23,560,800</td>
    <td align="right">2.9%</td>
    <td align="right">28.5%</td>
    <td align="left">No Conflicts</td>
    </tr>
    <tr>
    <td align="left">_START_EXECUTOR _SET_IP</td>
    <td align="right">21,741,760</td>
    <td align="right">2.6%</td>
    <td align="right">31.1%</td>
    <td align="left">Conflict. Both Use: Operand</td>
    </tr>
    <tr>
    <td align="left">_SET_IP _LOAD_ATTR</td>
    <td align="right">21,680,480</td>
    <td align="right">2.6%</td>
    <td align="right">33.7%</td>
    <td align="left">No Conflicts</td>
    </tr>
    <tr>
    <td align="left">_RESUME_CHECK _LOAD_FAST_0</td>
    <td align="right">21,058,400</td>
    <td align="right">2.6%</td>
    <td align="right">36.3%</td>
    <td align="left">No Conflicts</td>
    </tr>
    <tr>
    <td align="left">_CHECK_FUNCTION_EXACT_ARGS _CHECK_STACK_SPACE_OPERAND</td>
    <td align="right">19,557,660</td>
    <td align="right">2.4%</td>
    <td align="right">38.7%</td>
    <td align="left">Conflict. Both Use: Operand,Target</td>
    </tr>
    <tr>
    <td align="left">_INIT_CALL_PY_EXACT_ARGS_0 _SAVE_RETURN_OFFSET</td>
    <td align="right">19,042,580</td>
    <td align="right">2.3%</td>
    <td align="right">41.0%</td>
    <td align="left">No Conflicts</td>
    </tr>
    <tr>
    <td align="left">...</td>
    <td align="right">...</td>
    <td align="right">...</td>
    <td align="right">...</td>
    <td align="left">...</td>
    </tr>
</tbody>
</table>

* Issue: gh-119352

### CODE REVIEW

File: Tools/cases_generator/opcode_metadata_generator.py [hunk 1]
```diff
@@ -55,6 +55,7 @@
     "PASSTHROUGH",
     "OPARG_AND_1",
     "ERROR_NO_POP",
+    "OPERAND"
```
Comment (gvanrossum) [inline Tools/cases_generator/opcode_metadata_generator.py]: Trailing comma.
```suggestion
    "OPERAND",
```

File: Tools/scripts/summarize_stats.py [hunk 2]
```diff
@@ -1441,6 +1498,5 @@ def main():
 
     output_stats(args.inputs, json_output=args.json_output)
 
-
```
Comment (gvanrossum) [inline Tools/scripts/summarize_stats.py]: Keep it, please.

**Review by gvanrossum — COMMENTED** [review]
I started reviewing this but realized I don't know the summarize_stats.py script well enough to review that -- I think it's probably up to @mdboom to review that.

The cases_generator changes and the new generated output LGTM.

File: Tools/scripts/summarize_stats.py [hunk 3]
```diff
@@ -62,7 +62,7 @@ def get_defines(filepath: Path, prefix: str = "SPEC_FAIL"):
                     continue
                 line = line[len(start) :]
                 name, val = line.split()
-                defines[int(val.strip())].append(name.strip())
+                defines[int(val.strip().strip("()"))].append(name.strip())
```
Comment (mdboom) [inline Tools/scripts/summarize_stats.py:68]:
```suggestion
                defines[int(val.strip().removesuffix("()"))].append(name.strip())
```

This is probably safer, unless there is a need to remove things like `(())` that I'm missing.

File: Tools/scripts/summarize_stats.py [hunk 4]
```diff
@@ -120,6 +123,26 @@ def load_raw_data(input: Path) -> RawData:
 def save_raw_data(data: RawData, json_output: TextIO):
     json.dump(data, json_output)
 
+@functools.cache
+def _get_uop_flags_from_file(
+    flag_names: tuple[str] = None,
+    filepath: str | Path = "Include/internal/pycore_uop_metadata.h",
```
Comment (mdboom) [inline Tools/scripts/summarize_stats.py]:
```suggestion
    filepath: str | Path = Path("Include") / "internal" / "pycore_uop_metadata.h",
```

File: Tools/scripts/summarize_stats.py [hunk 5]
```diff
@@ -120,6 +123,26 @@ def load_raw_data(input: Path) -> RawData:
...
+) -> dict[str, list[str]]:
+    flags = {}
+    with open(SOURCE_DIR / filepath) as spec_src:
+        for line in spec_src:
+            if line.startswith("    [_"):
```
Comment (mdboom) [inline Tools/scripts/summarize_stats.py]: If the indentation in that file changes, we might silently not find anything in the file.  Maybe use a regular expression to match `"\s*\[_"` instead.

File: Tools/scripts/summarize_stats.py [hunk 6]
```diff
@@ -120,6 +123,26 @@ def load_raw_data(input: Path) -> RawData:
...
+                if name :=  re.search(r"\[[_A-Z0-9]+\]", line):
```
Comment (mdboom) [inline Tools/scripts/summarize_stats.py]:
```suggestion
                if name := re.search(r"\[[_A-Z0-9]+\]", line):
```

Formatting nit.  Also, this file should format cleanly with `black` (though our CI doesn't enforce it).  That's probably the easiest way to get the formatting consistent with the rest of the file.

File: Tools/scripts/summarize_stats.py [hunk 7]
```diff
@@ -120,6 +123,26 @@ def load_raw_data(input: Path) -> RawData:
...
+                    uop = name.group().strip("[]")
```
Comment (mdboom) [inline Tools/scripts/summarize_stats.py]:
```suggestion
                    uop = name.group().removesuffix("[]")
```

File: Tools/scripts/summarize_stats.py [hunk 8]
```diff
@@ -120,6 +123,26 @@ def load_raw_data(input: Path) -> RawData:
 def save_raw_data(data: RawData, json_output: TextIO):
     json.dump(data, json_output)
 
+@functools.cache
+def _get_uop_flags_from_file(
+    flag_names: tuple[str] = None,
+    filepath: str | Path = "Include/internal/pycore_uop_metadata.h",
+) -> dict[str, list[str]]:
+    flags = {}
```
Comment (mdboom) [inline Tools/scripts/summarize_stats.py:148]: Overall, I found this function a bit hard to follow (but then again, parsing C with regexes is always that way).  It might be helpful to have a header comment with an example of the pattern it's looking for and what the output would look like for it.

File: Tools/scripts/summarize_stats.py [hunk 9]
```diff
@@ -735,8 +758,32 @@ def execution_count_section() -> Section:
...
+        flag_compatible(
+            "HAS_OPERAND_FLAG",
+        ),
+        flag_compatible("HAS_JUMP_FLAG", "HAS_EXIT_FLAG", "HAS_DEOPT_FLAG"),
+    )
+    result_names = ("Oparg", "Operand", "Target")
+
+    if results.count(False) == 0:
+        return "No Conflict"
+    return f"Conflict. Both Use: {','.join(result_names[r] for r in range(3) if not results[r])}"
```
Comment (mdboom) [inline Tools/scripts/summarize_stats.py]: I think putting the two tuples in a dictionary makes working on them together a little easier.  (A tuple of tuples would probably also be fine here).

```suggestion
    results = {
        "Oparg": flag_compatible(
            "HAS_OPARG_AND_1_FLAG",
        ),
        "Operand": flag_compatible(
            "HAS_OPERAND_FLAG",
        ),
        "Target": flag_compatible("HAS_JUMP_FLAG", "HAS_EXIT_FLAG", "HAS_DEOPT_FLAG"),
    )
    
    if list(results.values()).count(False) == 0:
        return "No Conflict"
    return f"Conflict. Both Use: {','.join(k for k, v in results.items() if not v)}"
```

File: Tools/scripts/summarize_stats.py [hunk 10]
```diff
@@ -748,22 +795,32 @@ def calc_pair_count_table(stats: Stats) -> Rows:
...
+                Ratio(cumulative, total),
+            ]
+            if compat_data:
+                uop_flags = _get_uop_flags_from_file(
+                    tuple(v[0] for v in stats._data["_flag_defines"].values())
                 )
-            )
+                # for k, v in sorted(uop_flags.items()):
+                #    print(f"{k}:{v}")
+                # exit()
```
Comment (mdboom) [inline Tools/scripts/summarize_stats.py]: Maybe just debugging code left in?

**Review by mdboom — CHANGES_REQUESTED** [review]
This looks great, and thanks for working on this.  Just some stylistic changes mostly.  The example table in the PR was helpful to understand how this looks -- it's going to be super helpful.

**Comment by JeffersGlass** [comment]
Thanks for all the comments @mdboom! I’m taking a few days of post-conference R&R with my wife, but I will return to this PR presently and get it into shape.

File: Tools/scripts/summarize_stats.py (hunk 3 above)
Comment (JeffersGlass) [reply Tools/scripts/summarize_stats.py:68]: The defines for the flags look like `#define HAS_PASSTHROUGH_FLAG (4096)` - this line is to strip the parens around the number. We could do `resmovesuffix(")")` and `removeprefix("(")`?

Could also do the whole thing with a regex if that's preferable, but it would be a little unweildy.

File: Tools/scripts/summarize_stats.py (hunk 6 above)
Comment (JeffersGlass) [reply Tools/scripts/summarize_stats.py]: Thank you! Applied back and will fix.

File: Tools/scripts/summarize_stats.py (hunk 7 above)
Comment (JeffersGlass) [reply Tools/scripts/summarize_stats.py]: These lines look like `    [_STORE_SUBSCR_LIST_INT] = HAS_DEOPT_FLAG,` - this is to strip the [] around the name of the uop. Perhaps I'll just change the regex to only capture the Uop name in the first place.

Comment (JeffersGlass) [reply Tools/scripts/summarize_stats.py]: Change made in e9fc84d5.

File: Tools/scripts/summarize_stats.py (hunk 8 above)
Comment (JeffersGlass) [reply Tools/scripts/summarize_stats.py:148]: Agreed, I've added a comment to clarify. In general, I wonder if having something like `opcode_metadata_generator.py` for the UOp metadata, and spitting out that metadata as Python originally, would be more robust than parsing the C headers, but that might be a significant chunk of work.

File: Tools/scripts/summarize_stats.py (hunk 9 above)
Comment (JeffersGlass) [reply Tools/scripts/summarize_stats.py]: Thank you! Agreed, this is definitely clearere.

File: Tools/scripts/summarize_stats.py (hunk 10 above)
Comment (JeffersGlass) [reply Tools/scripts/summarize_stats.py]: Thanks!

File: Tools/scripts/summarize_stats.py (hunk 5 above)
Comment (JeffersGlass) [reply Tools/scripts/summarize_stats.py]: Makes sense, updated.

**Comment by JeffersGlass** [comment]
Does this want a NEWS.d entry? No worries either way, just wasn't sure whether Pystats updates in general (or this in particular) merits one.

File: Tools/scripts/summarize_stats.py (hunk 8 above)
Comment (mdboom) [reply Tools/scripts/summarize_stats.py:148]: I like that idea, but can/should be a follow-on to this PR.  I think parts of this predate having the generator -- it would be a lot cleaner to generate some JSON or something that would be easily parseable.

File: Tools/scripts/summarize_stats.py (hunk 3 above)
Comment (mdboom) [reply Tools/scripts/summarize_stats.py:68]: I see.  I misunderstood what this needed to do.  I think this is fine.

**Review by mdboom — APPROVED** [review]

**Comment by markshannon** [comment]
One minor request.
Can you shorten "Conflict. Both Use:" to just "Conflict:"? 
The "both" is redundant as we are talking about pairs, and it should result in less line wrapping.

**Comment by brandtbucher** [comment]
I think the ideal conflict calculation might be a bit trickier... our uop format is pretty subtle and confusing, since it's sort of grown organically. I think it deserves to be cleaned up at some point (but that's a separate issue).

The `target` member of a uop is defined as:

```c
union {
    uint32_t target;
    struct {
        union {
            uint16_t exit_index;
            uint16_t jump_target;
        };
        uint16_t error_target;
    };
};
```

- `target`: Only used by `_DEOPT` in the interpreter. No flag exists for checking this.
- `exit_index`: Only used by `_EXIT_TRACE` in the interpreter. No flag exists for checking this.
- `jump_target`: Used by instructions flagged as either `HAS_DEOPT_FLAG` or `HAS_EXIT_FLAG`
- `error_target`: Used by instructions flagged as either `HAS_ERROR_FLAG` or `HAS_ERROR_NO_POP_FLAG`

So I think it would be more like the following:

```py
targets = names.count("_DEOPT")
exit_indexes = names.count("_EXIT_TRACE")
jump_targets = combined_flags.count("HAS_DEOPT_FLAG") + combined_flags.count("HAS_EXIT_FLAG")
error_target = combined_flags.count("HAS_ERROR_FLAG") + combined_flags.count("HAS_ERROR_NO_POP_FLAG")

valid = (
    # UOP_FORMAT_TARGET:
    (targets <= 1 and exit_indexes == 0 and jump_targets == 0 and error_target == 0) or
    # UOP_FORMAT_EXIT:
    (targets == 0 and exit_indexes <= 1 and jump_targets == 0 and error_target <= 1) or
    # UOP_FORMAT_JUMP:
    (targets == 0 and exit_indexes == 0 and jump_targets <= 1 and error_target <= 1) or
)

```

**Comment by JeffersGlass** [comment]
Thanks @brandtbucher, I've updated the logic for detecting conflicting uses of of target.

Additionally - I had been detecting the use of an oparg based on solely `HAS_OPARG_AND_1_FLAG`. But it should it also include the use of `HAS_ARG_FLAG`, yes?

On a pyperformance pystats run, [none of the top 5000 pairs](https://gist.github.com/JeffersGlass/ce7ed9be0ac23d4940ab75eec50ae370) show a conflict for `target`... is that to be expected? It seems fishy to me, but I don't think I have enough intuition about the use of Target to be sure.

Here's a sample from the middle of those results:

<table>
<thead>
<tr>
<th align="left">Pair</th>
<th align="right">Count</th>
<th align="right">Self</th>
<th align="right">Cumulative</th>
<th align="left">Compatibility</th>
</tr>
</thead>
<tbody>
<td align="left">_CHECK_VALIDITY _LOAD_FAST</td>
<td align="right">9,082,739</td>
<td align="right">0.5%</td>
<td align="right">40.7%</td>
<td align="left">No Conflict</td>
</tr>
<tr>
<td align="left">_LOAD_FAST_4 _SET_IP</td>
<td align="right">8,849,139</td>
<td align="right">0.5%</td>
<td align="right">41.1%</td>
<td align="left">No Conflict</td>
</tr>
<tr>
<td align="left">_LOAD_CONST_INLINE_WITH_NULL _LOAD_FAST_5</td>
<td align="right">8,551,561</td>
<td align="right">0.5%</td>
<td align="right">41.6%</td>
<td align="left">No Conflict</td>
</tr>
<tr>
<td align="left">_CHECK_FUNCTION_EXACT_ARGS _CHECK_STACK_SPACE_OPERAND</td>
<td align="right">8,470,871</td>
<td align="right">0.5%</td>
<td align="right">42.1%</td>
<td align="left">Conflict: Operand</td>
</tr>
<tr>
<td align="left">_LOAD_FAST _LOAD_FAST</td>
<td align="right">8,465,883</td>
<td align="right">0.5%</td>
<td align="right">42.5%</td>
<td align="left">Conflict: Oparg</td>
</tr>
<tr>
<td align="left">_SET_IP _LOAD_ATTR</td>
<td align="right">8,337,465</td>
<td align="right">0.5%</td>
<td align="right">43.0%</td>
<td align="left">No Conflict</td>
</tr>
<tr>
<td align="left">_LOAD_FAST_3 _LOAD_FAST_4</td>
<td align="right">8,328,587</td>
<td align="right">0.5%</td>
<td align="right">43.4%</td>
<td align="left">No Conflict</td>
</tr>
<tr>
<td align="left">_GUARD_NOT_EXHAUSTED_LIST _ITER_NEXT_LIST</td>
<td align="right">8,311,545</td>
<td align="right">0.5%</td>
<td align="right">43.9%</td>
<td align="left">No Conflict</td>
</tr>
<tr>
<td align="left">_SET_IP _BINARY_OP</td>
<td align="right">8,137,465</td>
<td align="right">0.4%</td>
<td align="right">44.3%</td>
<td align="left">No Conflict</td>
</tr>
<tr>
<td align="left">_STORE_FAST _LOAD_FAST</td>
<td align="right">8,122,957</td>
<td align="right">0.4%</td>
<td align="right">44.8%</td>
<td align="left">Conflict: Oparg</td>
</tr>
<tr>
<td align="left">_LOAD_ATTR _CHECK_VALIDITY</td>
<td align="right">8,113,661</td>
<td align="right">0.4%</td>
<td align="right">45.2%</td>
<td align="left">No Conflict</td>
</tr>
<tr>
<td align="left">_RESUME_CHECK _LOAD_FAST_0</td>
<td align="right">8,094,103</td>
<td align="right">0.4%</td>
<td align="right">45.6%</td>
<td align="left">No Conflict</td>
</tr>
<tr>
<td align="left">_LOAD_CONST_INLINE_BORROW _LOAD_CONST_INLINE_BORROW</td>
<td align="right">8,012,851</td>
<td align="right">0.4%</td>
<td align="right">46.1%</td>
<td align="left">Conflict: Operand</td>
</tr>
<tr>
<td align="left">_GUARD_TYPE_VERSION _LOAD_ATTR_SLOT_0</td>
<td align="right">7,876,796</td>
<td align="right">0.4%</td>
<td align="right">46.5%</td>
<td align="left">Conflict: Operand</td>
</tr>
<tr>
<td align="left">...</td>
<td align="right">...</td>
<td align="right">...</td>
<td align="right">...</td>
<td align="left">...</td>
</tr>
</tbody>
</table>

**Comment by JeffersGlass** [comment]
Update: I do think the lack of target conflicts is a bug... I am troubleshooting/adding tests.

**Comment by JeffersGlass** [comment]
Indeed, I had an error in how I was combining the flags of two uops. Now corrected, the output looks something like this:

<table>
<thead>
<tr>
<th align="left">Pair</th>
<th align="right">Count</th>
<th align="right">Self</th>
<th align="right">Cumulative</th>
<th align="left">Compatibility</th>
</tr>
</thead>
<tbody>
<tr>
<td align="left">...</td>
<td align="right">...</td>
<td align="right">...</td>
<td align="right">...</td>
<td align="left">...</td>
</tr>
<tr>
<td align="left">_LOAD_FAST_3 _LOAD_FAST_4</td>
<td align="right">8,328,587</td>
<td align="right">0.5%</td>
<td align="right">43.4%</td>
<td align="left">No Conflict</td>
</tr>
<tr>
<td align="left">_GUARD_NOT_EXHAUSTED_LIST _ITER_NEXT_LIST</td>
<td align="right">8,311,545</td>
<td align="right">0.5%</td>
<td align="right">43.9%</td>
<td align="left">No Conflict</td>
</tr>
<tr>
<td align="left">_SET_IP _BINARY_OP</td>
<td align="right">8,137,465</td>
<td align="right">0.4%</td>
<td align="right">44.3%</td>
<td align="left">No Conflict</td>
</tr>
<tr>
<td align="left">_STORE_FAST _LOAD_FAST</td>
<td align="right">8,122,957</td>
<td align="right">0.4%</td>
<td align="right">44.8%</td>
<td align="left">Conflict: Oparg</td>
</tr>
<tr>
<td align="left">_LOAD_ATTR _CHECK_VALIDITY</td>
<td align="right">8,113,661</td>
<td align="right">0.4%</td>
<td align="right">45.2%</td>
<td align="left">No Conflict</td>
</tr>
<tr>
<td align="left">_RESUME_CHECK _LOAD_FAST_0</td>
<td align="right">8,094,103</td>
<td align="right">0.4%</td>
<td align="right">45.6%</td>
<td align="left">No Conflict</td>
</tr>
<tr>
<td align="left">_LOAD_CONST_INLINE_BORROW _LOAD_CONST_INLINE_BORROW</td>
<td align="right">8,012,851</td>
<td align="right">0.4%</td>
<td align="right">46.1%</td>
<td align="left">Conflict: Operand</td>
</tr>
<tr>
<td align="left">_GUARD_TYPE_VERSION _LOAD_ATTR_SLOT_0</td>
<td align="right">7,876,796</td>
<td align="right">0.4%</td>
<td align="right">46.5%</td>
<td align="left">Conflict: Operand,Target</td>
</tr>
<tr>
<td align="left">_CHECK_VALIDITY _LOAD_CONST_INLINE_BORROW</td>
<td align="right">7,720,714</td>
<td align="right">0.4%</td>
<td align="right">46.9%</td>
<td align="left">No Conflict</td>
</tr>
<tr>
<td align="left">_LOAD_FAST_2 _LOAD_FAST_3</td>
<td align="right">7,646,188</td>
<td align="right">0.4%</td>
<td align="right">47.3%</td>
<td align="left">No Conflict</td>
</tr>
<tr>
<td align="left">_TO_BOOL_BOOL _GUARD_IS_TRUE_POP</td>
<td align="right">7,517,743</td>
<td align="right">0.4%</td>
<td align="right">47.7%</td>
<td align="left">Conflict: Target</td>
</tr>
<tr>
<td align="left">_LOAD_FAST_2 _SET_IP</td>
<td align="right">7,374,642</td>
<td align="right">0.4%</td>
<td align="right">48.1%</td>
<td align="left">No Conflict</td>
</tr>
<tr>
<td align="left">_GUARD_TYPE_VERSION _LOAD_ATTR_METHOD_NO_DICT</td>
<td align="right">7,202,967</td>
<td align="right">0.4%</td>
<td align="right">48.5%</td>
<td align="left">Conflict: Operand</td>
</tr>
<tr>
<td align="left">...</td>
<td align="right">...</td>
<td align="right">...</td>
<td align="right">...</td>
<td align="left">...</td>
</tr>
</tbody>
</table>

**Comment by brandtbucher** [comment]
> Additionally - I had been detecting the use of an oparg based on solely `HAS_OPARG_AND_1_FLAG`. But it should it also include the use of `HAS_ARG_FLAG`, yes?

I *believe* `HAS_OPARG_AND_1_FLAG` implies `HAS_ARG_FLAG`, so only the latter should be necessary (in fact, I'm pretty sure that the versions of instructions with `HAS_OPARG_AND_1_FLAG` don't actually make it into the tier two code that's actually run, since they're replaced with other versions by the optimizer).

File: Tools/scripts/summarize_stats.py [hunk 11]
```diff
@@ -728,15 +771,76 @@ def execution_count_section() -> Section:
...
+
+    def flag_compatible(*flags: str):
+        return not any(
+            f in uop_flags[opcode_i] and f in uop_flags[opcode_j] for f in flags
+        )
+
+    def target_compatible(*opcodes: str):
+        targets = opcodes.count("_DEOPT")
+        exit_indexes = opcodes.count("_EXIT_TRACE")
+        combined_flags = list(itertools.chain(*(uop_flags[op] for op in opcodes)))
```
Comment (brandtbucher) [inline Tools/scripts/summarize_stats.py]:
```suggestion
        combined_flags = list(itertools.chain.from_iterable(uop_flags[op] for op in opcodes))
```

**Comment by JeffersGlass** [comment]
> I _believe_ `HAS_OPARG_AND_1_FLAG` implies `HAS_ARG_FLAG`, so only the latter should be necessary (in fact, I'm pretty sure that the versions of instructions with `HAS_OPARG_AND_1_FLAG` don't actually make it into the tier two code that's actually run, since they're replaced with other versions by the optimizer).

It does look like both of those things are true - I've adjusted the conflict-checking code to only account for `HAS_OPARG_FLAG`.

**Comment by JeffersGlass** [comment]
Give that we're [no longer pursuing superinstructions as a high-priority optimization](https://github.com/faster-cpython/ideas/issues/647), this is no longer going to be particularly helpful. I'll close this, though it could be revived if it turns out to be useful later.
